File: Objc/OCTCustomUITextViewLinkColor/OCTCustomUITextViewLinkColor/OCTTextView.m
//
//  OCTTextView.m
//  OCTCustomUITextViewLinkColor
//
//  Created by dmitry.brovkin on 4/26/17.
//  Copyright © 2017 dmitry.brovkin. All rights reserved.
//

#import "OCTTextView.h"

NSString *const OCTLinkAttributeName = @"OCTLinkAttributeName";


@implementation OCTTextView
{
    NSArray<NSString *> *_linksAttributes;
}

- (void)awakeFromNib {
    [super awakeFromNib];
    _linksAttributes = @[OCTLinkAttributeName, NSLinkAttributeName];
    
    UITapGestureRecognizer *tapGest = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(onTapAction:)];
    [self addGestureRecognizer:tapGest];
}

- (void)onTapAction:(UITapGestureRecognizer *)tapGest {
    CGPoint location = [tapGest locationInView:self];
    NSInteger charIndex = [self.layoutManager characterIndexForPoint:location
                                                     inTextContainer:self.textContainer
                            fractionOfDistanceBetweenInsertionPoints:nil];
    
    if (charIndex < self.textStorage.length) {
        NSRange range = NSMakeRange(0, 0);
        
        for (NSString *linkAttribute in _linksAttributes) {
            NSString *link = [self.attributedText attribute:linkAttribute atIndex:charIndex effectiveRange:&range];
            
            if ([link isKindOfClass:[NSString class]] && [link length]) {
                if (self.delegate && [self.delegate respondsToSelector:@selector(textView:shouldInteractWithURL:inRange:interaction:)]) {
                    [self.delegate textView:self shouldInteractWithURL:[NSURL URLWithString:link] inRange:range interaction:UITextItemInteractionInvokeDefaultAction];
                }
            }
        }
    }
}

@end

File: Swift/OCTCustomUITextViewLinkColor/OCTCustomUITextViewLinkColor/OCTTextView.swift
//
//  OCTTextView.swift
//  OCTCustomUITextViewLinkColor
//
//  Created by fantom on 4/20/17.
//  Copyright © 2017 dmitry.brovkin. All rights reserved.
//

import UIKit

let OCTLinkAttributeName = "OCTLinkAttributeName"

class OCTTextView: UITextView {
    
    private let _linksAttributes = [OCTLinkAttributeName, NSLinkAttributeName]
    
    override func awakeFromNib() {
        super.awakeFromNib()
        
        let tapGest = UITapGestureRecognizer(target: self, action: #selector(self.onTapAction))
        self.addGestureRecognizer(tapGest)
    }
    
    @objc private func onTapAction(_ tapGest: UITapGestureRecognizer) {
        let location = tapGest.location(in: self)
        let charIndex = self.layoutManager.characterIndex(for: location, in: self.textContainer, fractionOfDistanceBetweenInsertionPoints: nil)
        
        if charIndex < self.textStorage.length {
            var range = NSMakeRange(0, 0)
            
            for linkAttribute in _linksAttributes {
                if let link = self.attributedText.attribute(linkAttribute, at: charIndex, effectiveRange: &range) as? String {
                    _ = self.delegate?.textView?(self, shouldInteractWith: URL(string: link)!, in: range, interaction: .invokeDefaultAction)
                }
            }
        }
    }
}
